### PR TITLE
Impl partialeq for pylong

### PR DESCRIPTION
**Pull request by WilliamTakeshi**
Thanks for the great work on PyO3!

This PR implements `PartialEq` for `PyLong` with `u8`, `u16`, `u32`, `u64`, `u128`, `usize`, `i8`, `i16`, `i32`, `i64`, `i128` and `isize`. See #4250 

As this is my first contribution to the project, I welcome any feedback or suggestions.

### CODE REVIEW

File: src/types/num.rs
Comment (mejrs) [inline src/types/num.rs:29]: I don't think we should have this, at least not in this form. See https://github.com/PyO3/pyo3/issues/2221 for the prior discussion on this function.

File: src/types/num.rs
Comment (mejrs) [inline src/types/num.rs:49]: This implementation is incorrect if its value exceeds what `c_long` can hold (which is only an i32 on Windows). Also, I don't think having this implementation is a good idea.

File: src/types/num.rs
Comment (mejrs) [inline src/types/num.rs:152]: This and all the following methods will have different behavior on Windows and not Windows systems (because the size of `c_long` is different)

**Review by mejrs — CHANGES_REQUESTED** [review]
Thanks for the PR.

However I identify a couple of problems with it. You aren't handling overflow anywhere, which means unequal values van compare equal if overflow (on either side) occurs.

I think it would be better to just extract the PyLong to the compared-to integer type and compare that, rather than casting to `c_long`.

Finally, you can simplify this with a macro. For example:
```rust
macro_rules! int_compare {
    ($rust_type: ty) => {
        impl PartialEq<Bound<'_, PyLong>> for $rust_type {
            #[inline]
            fn eq(&self, other: &Bound<'_, PyLong>) -> bool {
                todo!()
            }
        }
    };
}

int_compare!(i8);
int_compare!(u8);
int_compare!(i16);
//etc
```

File: src/types/num.rs
Comment (WilliamTakeshi) [reply src/types/num.rs:49]: When you say its not a good idea to have this implementation, you mean having the `.value()` method?

I based my implementation heavily on `PyFloat`, which is why I wanted to clarify.

File: src/types/num.rs
Comment (WilliamTakeshi) [reply src/types/num.rs:29]: Thank you for the feedback! I'm still learning Rust, I greatly appreciate any help!

Before implementing `PartialEq`, I believe it's necessary to define a the constructor. In this regard, I have two potential approaches in mind: using the `num` crate to create a generic `new_bound` function, or defining an internal trait specific to this file.

What would you recommend as the best approach for this situation?

Comment (mejrs) [reply src/types/num.rs:29]: > Before implementing PartialEq, I believe it's necessary to define a the constructor.

How come? You dont need to create them to compare them. Anyway, the current way to create a PyLong is through `IntoPy`, so (strictly speaking) this function is not needed.

File: src/types/num.rs
Comment (mejrs) [reply src/types/num.rs:49]: It makes sense for PyFloat to have this method, because all possible values of a PyFloat can be expressed as a floating point number. But Python's integer has unlimited precision/size, so not all Python integers can be expressed with `c_long` (nor any other rust integer type like u64 or i128 for example).

File: src/types/num.rs
Comment (WilliamTakeshi) [reply src/types/num.rs]: Thanks! Removed the constructor!

File: src/types/num.rs
Comment (WilliamTakeshi) [reply src/types/num.rs]: Got it! I agree! Removed the value method!

File: src/types/num.rs
Comment (WilliamTakeshi) [reply src/types/num.rs]: Yes, definitely this cast wasn't safe for overflows. Improved to use `.extract()` instead

File: src/types/num.rs
Comment (davidhewitt) [inline src/types/num.rs:108]: Let's add some tests for overflows, e.g.

```suggestion
            assert_eq!(obj, v_usize);
            
            let big_num = (u8::MAX as u16) + 1;
            let big_obj = big_num.into_py(py).into_bound(py).downcast_into::<PyLong>().unwrap();
            
            for x in 0u8..+u8::MAX {
                assert_ne!(x, big_obj);
                assert_ne!(big_obj, x);
            }
```

**Review by davidhewitt — COMMENTED** [review]
Thanks, new implementation looks great. Just one further suggestion from me

File: tests/test_class_attributes.rs
Comment (davidhewitt) [inline tests/test_class_attributes.rs:227]: So just completing my read of the full diff, this is interesting. 

What I see here is that the introduction of these implementations has changed type inference here. It's a potentially breaking change. 

Is it a problem? I don't _think_ so; user code could already define other types implementing `FromPyObject` and `PartialEq<u8>` and thus have the same inference failure.

And I believe that this new `PartialEq` implementation here still holds potential convenience and usability improvements.

Comment (WilliamTakeshi) [reply tests/test_class_attributes.rs:227]: I see how this can be a breaking change. However, I don't think it's a major issue.

Ultimately, I'll leave the final decision to the maintainers haha :smiley: .

Comment (davidhewitt) [reply tests/test_class_attributes.rs:227]: Agreed, I think we move ahead with this.

**Review by davidhewitt — APPROVED** [review]
Looks great, thanks for this!

File: tests/test_class_attributes.rs
Comment (mejrs) [reply tests/test_class_attributes.rs:227]: I don't consider inference failures like this as breaking changes, so I'm also fine with this.

**Comment by davidhewitt** [comment]
I fixed the merge conflict and will proceed to merge, thanks again!

**Pull request merged**
